File: telemetry/aitf/sdk/typescript/src/exporters/ocsf-exporter.ts
/**
 * AITF OCSF Exporter.
 *
 * OTel SpanExporter that converts AI spans to OCSF events (reusing existing
 * OCSF classes) and exports them to SIEM/XDR endpoints, S3, or local files.
 *
 * Based on forwarder architecture from the AITelemetry project.
 */

import * as fs from "fs";
import * as path from "path";
import * as https from "https";
import { URL } from "url";
import { ReadableSpan, SpanExporter } from "@opentelemetry/sdk-trace-base";
import { ExportResult, ExportResultCode } from "@opentelemetry/core";
import { OCSFMapper } from "../ocsf/mapper";
import { ComplianceMapper } from "../ocsf/compliance-mapper";
import { AIBaseEvent, stripNulls } from "../ocsf/schema";

/** Localhost hosts allowed for HTTP (development only). */
const DEV_HOSTS = new Set(["localhost", "127.0.0.1", "::1"]);

/** Options for configuring the OCSFExporter. */
export interface OCSFExporterOptions {
  endpoint?: string;
  outputFile?: string;
  complianceFrameworks?: string[];
  includeRawSpan?: boolean;
  apiKey?: string;
}

/**
 * Validate endpoint URL for security.
 * Enforces HTTPS when API key is present and not localhost.
 */
function validateEndpoint(endpoint: string, apiKey?: string): void {
  const url = new URL(endpoint);

  if (url.protocol !== "http:" && url.protocol !== "https:") {
    throw new Error(
      `Unsupported URL scheme '${url.protocol}'. Only http: and https: are allowed.`
    );
  }

  const isDev = DEV_HOSTS.has(url.hostname);

  // Enforce HTTPS when API key is present and not localhost
  if (apiKey && url.protocol !== "https:" && !isDev) {
    throw new Error(
      "HTTPS is required when using API key authentication. " +
        "Use https:// or connect to localhost for development."
    );
  }
}

/**
 * Validate output file path to prevent path traversal.
 */
function validateOutputPath(outputFile: string): string {
  if (outputFile.includes("..")) {
    throw new Error(
      `Path traversal detected in output path: ${outputFile}`
    );
  }
  return path.resolve(outputFile);
}

/**
 * Exports OTel spans as OCSF AI events (reusing existing OCSF classes).
 *
 * Converts AI-related spans to OCSF events using OCSFMapper,
 * enriches with compliance metadata, and exports to configured
 * destinations.
 *
 * Usage:
 *   import { NodeTracerProvider } from "@opentelemetry/sdk-trace-node";
 *   import { BatchSpanProcessor } from "@opentelemetry/sdk-trace-base";
 *   import { OCSFExporter } from "@aitf/sdk";
 *
 *   const exporter = new OCSFExporter({
 *     outputFile: "/var/log/aitf/ocsf_events.jsonl",
 *     complianceFrameworks: ["nist_ai_rmf", "eu_ai_act", "mitre_atlas"],
 *   });
 *   const provider = new NodeTracerProvider();
 *   provider.addSpanProcessor(new BatchSpanProcessor(exporter));
 */
export class OCSFExporter implements SpanExporter {
  private readonly _endpoint: string | null;
  private readonly _outputFile: string | null;
  private readonly _includeRawSpan: boolean;
  private readonly _apiKey: string | null;
  private readonly _mapper: OCSFMapper;
  private readonly _complianceMapper: ComplianceMapper;
  private _eventCount = 0;

  constructor(options: OCSFExporterOptions = {}) {
    this._apiKey = options.apiKey ?? null;
    this._includeRawSpan = options.includeRawSpan ?? false;
    this._mapper = new OCSFMapper();
    this._complianceMapper = new ComplianceMapper({
      frameworks: options.complianceFrameworks,
    });

    // Validate endpoint URL
    if (options.endpoint) {
      validateEndpoint(options.endpoint, options.apiKey);
      this._endpoint = options.endpoint;
    } else {
      this._endpoint = null;
    }

    // Validate and ensure output directory exists
    if (options.outputFile) {
      const resolved = validateOutputPath(options.outputFile);
      this._outputFile = resolved;
      const dir = path.dirname(resolved);
      fs.mkdirSync(dir, { recursive: true });
    } else {
      this._outputFile = null;
    }
  }

  /**
   * Export spans as OCSF events.
   */
  export(
    spans: ReadableSpan[],
    resultCallback: (result: ExportResult) => void
  ): void {
    const events: Record<string, unknown>[] = [];

    for (const span of spans) {
      const ocsfEvent = this._mapper.mapSpan(span);
      if (!ocsfEvent) {
        continue;
      }

      // Enrich with compliance metadata
      const eventType = this._classifyEvent(ocsfEvent);
      if (eventType) {
        this._complianceMapper.enrichEvent(ocsfEvent, eventType);
      }

      // Runtime validation before type cast to Record<string, unknown>
      if (
        typeof ocsfEvent !== "object" ||
        ocsfEvent === null ||
        Array.isArray(ocsfEvent)
      ) {
        console.warn("OCSF: Skipping non-object event from mapper");
        continue;
      }
      const eventDict = stripNulls(
        ocsfEvent as unknown as Record<string, unknown>
      );
      events.push(eventDict);
      this._eventCount++;
    }

    if (events.length === 0) {
      resultCallback({ code: ExportResultCode.SUCCESS });
      return;
    }

    try {
      if (this._outputFile) {
        this._exportToFile(events);
      }
      if (this._endpoint) {
        this._exportToEndpoint(events)
          .then(() => {
            resultCallback({ code: ExportResultCode.SUCCESS });
          })
          .catch((err) => {
            console.error("OCSF export to endpoint failed:", this._sanitizeError(err));
            resultCallback({ code: ExportResultCode.FAILED });
          });
        return;
      }
      resultCallback({ code: ExportResultCode.SUCCESS });
    } catch (err) {
      console.error("OCSF export failed:", this._sanitizeError(err));
      resultCallback({ code: ExportResultCode.FAILED });
    }
  }

  private _exportToFile(events: Record<string, unknown>[]): void {
    if (!this._outputFile) return;

    const lines = events
      .map((event) => JSON.stringify(event))
      .join("\n");
    fs.appendFileSync(this._outputFile, lines + "\n", {
      encoding: "utf-8",
      mode: 0o600, // Restrictive file permissions
    });
  }

  private async _exportToEndpoint(
    events: Record<string, unknown>[]
  ): Promise<void> {
    if (!this._endpoint) return;

    const url = new URL(this._endpoint);
    const payload = JSON.stringify(events);

    // Enforce HTTPS for all non-localhost endpoints
    if (url.protocol !== "https:" && !DEV_HOSTS.has(url.hostname)) {
      console.warn(
        "OCSF: Using insecure HTTP endpoint. Consider using HTTPS for production."
      );
    }

    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      "Content-Length": String(Buffer.byteLength(payload)),
    };
    if (this._apiKey) {
      headers["Authorization"] = `Bearer ${this._apiKey}`;
    }

    return new Promise<void>((resolve, reject) => {
      const requestOptions = {
        hostname: url.hostname,
        port: url.port || (url.protocol === "https:" ? 443 : 80),
        path: url.pathname + url.search,
        method: "POST",
        headers,
        timeout: 30000,
      };

      const req = https.request(requestOptions, (res) => {
        if (res.statusCode && res.statusCode >= 400) {
          reject(
            new Error(
              `OCSF endpoint returned ${res.statusCode}`
            )
          );
        } else {
          resolve();
        }
        // Consume response data to free memory
        res.resume();
      });

      req.on("error", reject);
      req.on("timeout", () => {
        req.destroy();
        reject(new Error("OCSF endpoint request timed out"));
      });

      req.write(payload);
      req.end();
    });
  }

  /**
   * Sanitize an error before logging to ensure the API key is never leaked.
   */
  private _sanitizeError(err: unknown): string {
    const message = err instanceof Error ? err.message : String(err);
    if (this._apiKey) {
      return message.replaceAll(this._apiKey, "[REDACTED]");
    }
    return message;
  }

  /**
   * Classify an OCSF event for compliance mapping.
   *
   * AITF reuses existing OCSF classes (OCSF PR #1641 / issue #1640), so
   * `class_uid` is no longer unique per AI event type (e.g. inference and
   * tool execution both reuse API Activity 6003). Classify by the
   * discriminating fields the mapper sets on each event.
   */
  private _classifyEvent(event: AIBaseEvent): string | null {
    const e = event as unknown as Record<string, unknown>;
    if ("model" in e) return "model_inference";
    if ("tool_name" in e) return "tool_execution";
    if ("database_name" in e) return "data_retrieval";
    if ("finding" in e) return "security_finding";
    if ("session_id" in e && "agent_id" in e) return "agent_activity";
    return null;
  }

  /** Number of OCSF events exported. */
  get eventCount(): number {
    return this._eventCount;
  }

  async shutdown(): Promise<void> {
    // No-op
  }

  async forceFlush(): Promise<void> {
    // No-op
  }
}

File: telemetry/aitf/sdk/python/src/aitf/exporters/ocsf_exporter.py
"""AITF OCSF Exporter.

OTel SpanExporter that converts AI spans to OCSF events (class-reuse model)
and exports them to SIEM/XDR endpoints, S3, or local files.

Based on forwarder architecture from the AITelemetry project.
"""

from __future__ import annotations

import json
import logging
import ssl
import threading
from pathlib import Path
from typing import Any, Sequence
from urllib.parse import urlparse

from opentelemetry.sdk.trace import ReadableSpan
from opentelemetry.sdk.trace.export import SpanExporter, SpanExportResult

from aitf.ocsf.compliance_mapper import ComplianceMapper
from aitf.ocsf.mapper import OCSFMapper
from aitf.ocsf.schema import AIBaseEvent

logger = logging.getLogger(__name__)

# Maximum output file size (500MB) to prevent unbounded growth
_MAX_FILE_SIZE_BYTES = 500 * 1024 * 1024

# Localhost hosts allowed for HTTP (development only)
_DEV_HOSTS = {"localhost", "127.0.0.1", "::1"}


def _validate_endpoint(endpoint: str, api_key: str | None = None) -> str:
    """Validate and sanitize an endpoint URL.

    Enforces HTTPS for non-localhost endpoints, especially when API keys
    are used. Prevents SSRF by restricting to http(s) schemes.
    """
    parsed = urlparse(endpoint)

    if parsed.scheme not in ("http", "https"):
        raise ValueError(
            f"Unsupported URL scheme '{parsed.scheme}'. Only http and https are allowed."
        )

    is_dev_host = parsed.hostname in _DEV_HOSTS if parsed.hostname else False

    # Enforce HTTPS when API key is present and not localhost
    if api_key and parsed.scheme != "https" and not is_dev_host:
        raise ValueError(
            "HTTPS is required when using API key authentication. "
            "Use https:// or connect to localhost for development."
        )

    # Warn if using HTTP even without API key (non-localhost)
    if parsed.scheme == "http" and not is_dev_host:
        logger.warning(
            "Using insecure HTTP endpoint. Consider using HTTPS for production."
        )

    return endpoint


def _validate_output_path(output_file: str) -> Path:
    """Validate output file path to prevent path traversal.

    Rejects any path containing '..' components, which could be used
    to escape the intended directory.
    """
    raw_path = Path(output_file)

    # Reject paths with '..' components to prevent traversal
    if ".." in raw_path.parts:
        raise ValueError(
            f"Path traversal detected in output path: {output_file!r} "
            f"contains '..' component"
        )

    return raw_path.resolve()


class OCSFExporter(SpanExporter):
    """Exports OTel spans as OCSF AI events (class-reuse model).

    Converts AI-related spans to OCSF events using OCSFMapper,
    enriches with compliance metadata, and exports to configured
    destinations.

    Usage:
        from opentelemetry.sdk.trace import TracerProvider
        from opentelemetry.sdk.trace.export import BatchSpanProcessor

        exporter = OCSFExporter(
            output_file="/var/log/aitf/ocsf_events.jsonl",
            compliance_frameworks=["nist_ai_rmf", "eu_ai_act", "mitre_atlas"],
        )
        provider = TracerProvider()
        provider.add_span_processor(BatchSpanProcessor(exporter))
    """

    def __init__(
        self,
        endpoint: str | None = None,
        output_file: str | None = None,
        compliance_frameworks: list[str] | None = None,
        include_raw_span: bool = False,
        api_key: str | None = None,
    ):
        # Validate endpoint URL
        if endpoint:
            self._endpoint = _validate_endpoint(endpoint, api_key)
        else:
            self._endpoint = None

        self._include_raw_span = include_raw_span
        self._api_key = api_key
        self._mapper = OCSFMapper()
        self._compliance_mapper = ComplianceMapper(frameworks=compliance_frameworks)
        self._event_count = 0
        self._lock = threading.Lock()
        self._file_lock = threading.Lock()

        # Validate and create output file path
        if output_file:
            resolved = _validate_output_path(output_file)
            self._output_file = str(resolved)
            resolved.parent.mkdir(parents=True, exist_ok=True)
        else:
            self._output_file = None

    def export(self, spans: Sequence[ReadableSpan]) -> SpanExportResult:
        """Export spans as OCSF events."""
        events: list[dict[str, Any]] = []

        for span in spans:
            ocsf_event = self._mapper.map_span(span)
            if ocsf_event is None:
                continue

            # Enrich with compliance metadata
            event_type = self._classify_event(ocsf_event)
            if event_type:
                self._compliance_mapper.enrich_event(ocsf_event, event_type)

            event_dict = ocsf_event.model_dump(exclude_none=True)
            events.append(event_dict)
            with self._lock:
                self._event_count += 1

        if not events:
            return SpanExportResult.SUCCESS

        # Export to configured destinations
        try:
            if self._output_file:
                self._export_to_file(events)
            if self._endpoint:
                self._export_to_endpoint(events)
            return SpanExportResult.SUCCESS
        except Exception as exc:
            logger.error("OCSF export failed: %s", exc)
            return SpanExportResult.FAILURE

    def _export_to_file(self, events: list[dict[str, Any]]) -> None:
        """Write OCSF events to JSONL file with size limit enforcement."""
        with self._file_lock:
            output_path = Path(self._output_file)

            # Check file size to prevent unbounded growth
            if output_path.exists() and output_path.stat().st_size > _MAX_FILE_SIZE_BYTES:
                logger.warning(
                    "OCSF output file exceeds %d bytes, rotating",
                    _MAX_FILE_SIZE_BYTES,
                )
                rotated = output_path.with_suffix(".jsonl.old")
                output_path.rename(rotated)

            with open(self._output_file, "a") as f:
                for event in events:
                    f.write(json.dumps(event, default=str) + "\n")

    def _export_to_endpoint(self, events: list[dict[str, Any]]) -> None:
        """Send OCSF events to HTTP endpoint with TLS enforcement."""
        try:
            import urllib.request

            headers = {"Content-Type": "application/json"}
            if self._api_key:
                headers["Authorization"] = f"Bearer {self._api_key}"

            payload = json.dumps(events, default=str).encode("utf-8")
            req = urllib.request.Request(
                self._endpoint,
                data=payload,
                headers=headers,
                method="POST",
            )

            # Create SSL context for HTTPS connections with strict verification
            ssl_context = ssl.create_default_context()
            ssl_context.check_hostname = True
            ssl_context.verify_mode = ssl.CERT_REQUIRED

            with urllib.request.urlopen(req, timeout=30, context=ssl_context) as resp:
                if resp.status >= 400:
                    logger.error("OCSF endpoint returned %d", resp.status)
        except Exception as exc:
            logger.error("Failed to send to OCSF endpoint: %s", exc)
            raise

    def _classify_event(self, event: AIBaseEvent) -> str | None:
        """Classify an OCSF event for compliance mapping.

        AITF reuses existing OCSF classes (OCSF PR #1641 / issue #1640), so
        ``class_uid`` is no longer unique per AI event type (e.g. inference and
        tool execution both reuse API Activity 6003). Classify by event type.
        """
        from aitf.ocsf.event_classes import (
            AIAgentActivityEvent,
            AIAssetInventoryEvent,
            AIDataRetrievalEvent,
            AIGovernanceEvent,
            AIIdentityEvent,
            AIModelInferenceEvent,
            AIModelOpsEvent,
            AISecurityFindingEvent,
            AISupplyChainEvent,
            AIToolExecutionEvent,
        )

        type_map = [
            (AIModelInferenceEvent, "model_inference"),
            (AIAgentActivityEvent, "agent_activity"),
            (AIToolExecutionEvent, "tool_execution"),
            (AIDataRetrievalEvent, "data_retrieval"),
            (AISecurityFindingEvent, "security_finding"),
            (AISupplyChainEvent, "supply_chain"),
            (AIGovernanceEvent, "governance"),
            (AIIdentityEvent, "identity"),
            (AIModelOpsEvent, "model_ops"),
            (AIAssetInventoryEvent, "asset_inventory"),
        ]
        for cls, name in type_map:
            if isinstance(event, cls):
                return name
        return None

    @property
    def event_count(self) -> int:
        with self._lock:
            return self._event_count

    def shutdown(self) -> None:
        pass

    def force_flush(self, timeout_millis: int = 30000) -> bool:
        return True
